cache/cache.go

```diff
@@ -0,0 +1,120 @@
+package cache
+
+import (
+	"context"
+	"encoding"
+	"sync/atomic"
+	"time"
+)
+
+// Client defines a cache instance.
+//
+// This can represent the cache for the entire system or for a particular use-case/type.
+//
+// If a cache is used for multiple purposes, then care must be taken to ensure uniqueness of cache keys.
+//
+// It is not recommended to change this struct's member data after creation as a data race will likely ensue.
+type Client struct {
+	Storage Storage
+	Logger  Logger
+	Metrics Metrics
+
+	// track pending cache writes
+	pendingWrites uint64
+}
+
+// Get attempts to retrieve the value from cache and when it misses will run the builder func to create the value.
+//
+// It will asynchronously update/save the value in the cache on after a successful builder run
+func (c *Client) Get(ctx context.Context, key string, dest BinaryEncoder, builder Builder) error {
+	bytes, err := c.Storage.Get(ctx, key)
+	if err != nil {
+		if err == errCacheMiss {
+			c.getMetrics().Track(CacheMiss)
+
+			err = builder.Build(ctx, key, dest)
+			if err != nil {
+				c.getMetrics().Track(CacheLambdaError)
+				return err
+			}
+
+			atomic.AddUint64(&c.pendingWrites, 1)
+			go c.updateCache(key, dest)
+
+			return err
+		}
+
+		c.getMetrics().Track(CacheError)
+		return err
+	}
+
+	err = dest.UnmarshalBinary(bytes)
+	if err != nil {
+		c.getMetrics().Track(CacheUnmarshalError)
+		return err
+	}
+
+	c.getMetrics().Track(CacheHit)
+	return nil
+}
+
+// update the cache with the supplied key/value pair
+func (c *Client) updateCache(key string, val encoding.BinaryMarshaler) {
+	defer func() {
+		// update tracking
+		atomic.AddUint64(&c.pendingWrites, ^uint64(0))
```

Comment:
I would suggest to use `int64` for this counter because you won't have more than `9223372036854775807` concurrent writes I suppose 😄 

Moreover, `atomic.AddInt64(&c.pendingWrites, -1)` is more intuitive and easier to debug by looking for negative pendingWrites.